<a href="https://colab.research.google.com/github/Hirusha99/Regression/blob/main/Binary_Predictors_in_a_Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Binary Predictors in a Logistic Regression

Using the same code as in the previous exercise, find the odds of 'duration'.

What do they tell you?

## Import the relevant libraries

In [1]:
import pandas as pd
import  numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set

import statsmodels.api as sm

#apply a fix to the statsmodels Library
from scipy import stats
stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq,df)

## Load the data

Load the ‘Bank_data.csv’ dataset.

In [2]:
data = pd.read_csv('/content/008 Bank-data.csv')

In [3]:
data.head()

Unnamed: 0  interest_rate  credit  march  may  previous  duration    y
0           0          1.334     0.0    1.0  0.0       0.0     117.0   no
1           1          0.767     0.0    0.0  2.0       1.0     274.0  yes
2           2          4.858     0.0    1.0  0.0       0.0     167.0   no
3           3          4.120     0.0    0.0  0.0       0.0     686.0  yes
4           4          4.856     0.0    1.0  0.0       0.0     157.0   no

In [4]:
data.describe()

Unnamed: 0  interest_rate      credit       march         may  \
count  518.000000     518.000000  518.000000  518.000000  518.000000   
mean   258.500000       2.835776    0.034749    0.266409    0.388031   
std    149.677988       1.876903    0.183321    0.442508    0.814527   
min      0.000000       0.635000    0.000000    0.000000    0.000000   
25%    129.250000       1.042750    0.000000    0.000000    0.000000   
50%    258.500000       1.466000    0.000000    0.000000    0.000000   
75%    387.750000       4.956500    0.000000    1.000000    0.000000   
max    517.000000       4.970000    1.000000    1.000000    5.000000   

         previous     duration  
count  518.000000   518.000000  
mean     0.127413   382.177606  
std      0.333758   344.295990  
min      0.000000     9.000000  
25%      0.000000   155.000000  
50%      0.000000   266.500000  
75%      0.000000   482.750000  
max      1.000000  2653.000000

In [7]:
data['y'] = data['y'].map({'yes':1, 'no': 0})

In [8]:
data.head()

Unnamed: 0  interest_rate  credit  march  may  previous  duration  y
0           0          1.334     0.0    1.0  0.0       0.0     117.0  0
1           1          0.767     0.0    0.0  2.0       1.0     274.0  1
2           2          4.858     0.0    1.0  0.0       0.0     167.0  0
3           3          4.120     0.0    0.0  0.0       0.0     686.0  1
4           4          4.856     0.0    1.0  0.0       0.0     157.0  0

### Declare the dependent and independent variables

Use 'duration' as the independet variable.

In [11]:
x1 = data['duration']
y = data['y']

### Simple Logistic Regression

Run the regression.

In [13]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
result_log = reg_log.fit()
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.546118
         Iterations 7


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                  518
Model:                          Logit   Df Residuals:                      516
Method:                           MLE   Df Model:                            1
Date:                Tue, 26 Dec 2023   Pseudo R-squ.:                  0.2121
Time:                        04:46:08   Log-Likelihood:                -282.89
converged:                       True   LL-Null:                       -359.05
Covariance Type:            nonrobust   LLR p-value:                 5.387e-35
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.7001      0.192     -8.863      0.000      -2.076      -1.324
duration       0.0051      0.001      9.159      0.000       0.004       0.006
==============================================================================
"""

### Find the odds of duration

The odds of duration are the exponential of the log odds from the summary table

In [14]:
np.exp(0.0051)

1.005113027136717

# Accuracy

In [16]:
result_log.predict()

array([0.24936904, 0.42571348, 0.30019503, 0.85898342, 0.28956627,
       0.2580803 , 0.21914028, 0.16614452, 0.86976608, 0.31987434,
       0.55276776, 0.22532533, 0.30234734, 0.39480294, 0.35411888,
       0.489064  , 0.29167414, 0.25225084, 0.50567672, 0.82802726,
       0.84413091, 0.5890681 , 0.45848474, 0.40215497, 0.44961416,
       0.20546539, 0.4157464 , 0.42321571, 0.58286683, 0.2580803 ,
       0.17928751, 0.2580803 , 0.85711543, 0.43322907, 0.293791  ,
       0.32658401, 0.98373368, 0.21652721, 0.99665611, 0.62920727,
       0.2639969 , 0.49417474, 0.30885632, 0.27711089, 0.28746746,
       0.2474601 , 0.19643708, 0.30342677, 0.30559213, 0.4712014 ,
       0.87376626, 0.20132396, 0.22177596, 0.28642146, 0.25612742,
       0.49417474, 0.27916371, 0.20214723, 0.88872861, 0.53501481,
       0.23806331, 0.24556099, 0.90211616, 0.40092648, 0.51078751,
       0.45975425, 0.28019361, 0.30450837, 0.87432905, 0.23713725,
       0.18231618, 0.3706541 , 0.4724754 , 0.38144917, 0.54009

In [17]:
# predicted values
np.set_printoptions(formatter={'float': lambda x:"{0:0.2f}".format(x)})
result_log.predict()

array([0.25, 0.43, 0.30, 0.86, 0.29, 0.26, 0.22, 0.17, 0.87, 0.32, 0.55,
       0.23, 0.30, 0.39, 0.35, 0.49, 0.29, 0.25, 0.51, 0.83, 0.84, 0.59,
       0.46, 0.40, 0.45, 0.21, 0.42, 0.42, 0.58, 0.26, 0.18, 0.26, 0.86,
       0.43, 0.29, 0.33, 0.98, 0.22, 1.00, 0.63, 0.26, 0.49, 0.31, 0.28,
       0.29, 0.25, 0.20, 0.30, 0.31, 0.47, 0.87, 0.20, 0.22, 0.29, 0.26,
       0.49, 0.28, 0.20, 0.89, 0.54, 0.24, 0.25, 0.90, 0.40, 0.51, 0.46,
       0.28, 0.30, 0.87, 0.24, 0.18, 0.37, 0.47, 0.38, 0.54, 0.26, 0.41,
       0.57, 0.29, 0.21, 0.34, 0.43, 0.30, 0.46, 0.41, 0.97, 0.30, 0.53,
       0.40, 0.23, 0.36, 0.22, 0.37, 0.32, 0.77, 0.81, 0.80, 0.20, 0.98,
       0.30, 0.93, 0.34, 0.89, 0.41, 0.50, 0.50, 0.42, 0.35, 0.42, 0.52,
       0.54, 0.25, 0.31, 0.21, 0.74, 0.18, 0.25, 0.88, 0.60, 0.48, 0.23,
       0.17, 0.40, 0.99, 0.48, 0.96, 0.36, 0.59, 0.34, 0.42, 0.29, 0.33,
       0.48, 0.90, 0.46, 0.68, 0.57, 0.43, 0.43, 0.31, 0.26, 0.23, 0.84,
       0.44, 0.84, 0.65, 0.38, 0.26, 1.00, 0.26, 0.

In [18]:
# acctual  values
np.array(data['y'])

array([0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0,
       1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1,
       0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1,
       1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0,

In [19]:
# confusion matrix
result_log.pred_table()

array([[204.00, 55.00],
       [104.00, 155.00]])

In [21]:
df = pd.DataFrame(result_log.pred_table())
df.columns = ['predicted 0','predicted 1']
df = df.rename(index ={0: 'Actual 0', 1: 'Actual 1'})
df


predicted 0  predicted 1
Actual 0        204.0         55.0
Actual 1        104.0        155.0

for 204 observations the model predicted 0 and the true vale was 0

for 155 observations the model predicted 1 and the true vale was 1

for 55 observations the model predicted 1 and the true vale was 0

for 104 observations the model predicted 0 and the true vale was 1

In [22]:
# get the accuracy
cm = np.array(df)
accuracy = (cm[0,0]+cm[1,1])/cm.sum()
accuracy

0.693050193050193